pennylane/dla/lie_closure.py

```diff
@@ -0,0 +1,282 @@
+            # The input PauliSentence must have been linearly dependent
+            return M, pw_to_idx, rank, num_pw, False
+
+        M = np.pad(M, ((0, new_num_pw - num_pw), (0, 1)))
+
+        # If there are new PauliWords (i.e. new basis vectors), the candidate vector must be linearly independent
+        if new_num_pw > num_pw:
+            new_pw_to_idx = copy(pw_to_idx)
+            for i, pw in enumerate(new_pws, start=num_pw):
+                new_pw_to_idx[pw] = i
+            # Add new PauliSentence entries to matrix
+            for pw, value in pauli_sentence.items():
+                M[new_pw_to_idx[pw], rank] = value
+
+            return M, new_pw_to_idx, rank + 1, new_num_pw, True
+
+        # Add new PauliSentence entries to matrix
+        for pw, value in pauli_sentence.items():
+            M[pw_to_idx[pw], rank] = value
+
+        # Check if new vector is proportional to any of the previous vectors
+        # This is significantly cheaper than computing the rank and should be done first
+        if _is_any_col_propto_last(M):
+            M = M[:num_pw, :rank]
+            return M, pw_to_idx, rank, num_pw, False
+
+        new_rank = np.linalg.matrix_rank(M)  # expensive
+
+        # Manual singular value alternative, probably slower than ``matrix_rank``
+        # sing_value = np.min(np.abs(svd(M, compute_uv=False, lapack_driver="gesdd", check_finite=False)))
+        if new_rank == rank:
+            M = M[:num_pw, :rank]
+            return M, pw_to_idx, rank, num_pw, False
+
+        return M, pw_to_idx, rank + 1, new_num_pw, True
+
+    def __repr__(self):
+        return str(self.basis)
+
+    def __eq__(self, other):
+        """
+        Two VSpaces are equivalent when they span the same dimensional space.
+        This is checked here by having matching PauliWord keys in the sparse DOK representation and having the same rank.
+        """
+        if not self._num_pw == other._num_pw:
+            return False
+        if not set(self._pw_to_idx.keys()) == set(other._pw_to_idx.keys()):
+            return False
+
+        rank1 = np.linalg.matrix_rank(self._M)
+        rank2 = np.linalg.matrix_rank(other._M)
+        rank3 = np.linalg.matrix_rank(np.concatenate([self._M, other._M], axis=1))
+
+        return rank1 == rank2 and rank2 == rank3
+
+
+def _is_any_col_propto_last(inM):
+    """Given a 2D matrix M, check if any column is proportional to the last column
+    **Example**
+
+    .. code-block::python3
+        M1 = np.array([
+            [0., 1., 2., 4.],
+            [1., 1., 1., 0.],
+            [2., 2., 3., 6.]
+        ])
+        M2 = np.array([
+            [0., 1., 2., 4.],
+            [1., 1., 0., 0.],
+            [2., 2., 3., 6.]
+        ])
+
+    >>> _any_col_propto(M1)
+    False
+    >>> _any_col_propto(M2)
+    True
+
+    """
+    M = inM.copy()
+
+    nonzero_mask = np.nonzero(M[:, 0])  # target vector is the last column
```

Comment:
Doesn't this make the target vector the first column of `M`? For the last column it would be
```suggestion
    nonzero_mask = np.nonzero(M[:, -1])  # target vector is the last column
```